snakemake/satija/src/generate_cell_sets.py

```diff
@@ -0,0 +1,185 @@
+        assert CL_ROOT_ID in ancestor_term_set
+
+        # Get the parents of the current node.
+        node_parents = list(graph.out_edges(node_id, keys=True))
+
+        up_dag_paths = []
+        for node_parent in node_parents:
+            _, curr_parent_id, relationship = node_parent
+            if relationship == "is_a":
+                parent_paths = get_parents(curr_parent_id)
+                for parent_path in parent_paths:
+                    up_dag_paths.append([node_id] + parent_path)
+        return up_dag_paths
+
+    ancestors_and_sets = []
+
+    for cell_type, cell_type_df in df.groupby(COLUMNS.ANNOTATION.value):
+        try:
+            node_id = name_to_id[cell_type]
+        except KeyError:
+            print((
+                f"ERROR: annotation '{cell_type}' does "
+                "not match any node in the cell ontology."
+            ))
+            continue
+
+        # Get all of the possible paths up to the root
+        # from the current node.
+        paths_up = get_parents(node_id)
+        # Get the names of each node in each path.
+        named_paths_up = [
+            [id_to_name[n_id] for n_id in path_nodes]
+            for path_nodes in paths_up
+        ]
+        print((
+            f"WARNING: {id_to_name[node_id]} has {len(paths_up)} paths"
+            f" up to {CL_ROOT_ID} ({id_to_name[CL_ROOT_ID]})."
+        ))
+
+        # Sort potential paths "up the hierarchy" by our preferences,
+        # to avoid "functional" parent nodes like "motile cell"
+        sorted_named_paths_up = sort_paths_up_cell_ontology(named_paths_up)
+
+        named_ancestors = sorted_named_paths_up[0]
+        named_ancestors_reversed = list(reversed(named_ancestors))
+        # Get a list of (cell_id, prediction_score) tuples for the set.
+        set_value = [
+            list(x)
+            for x in zip(
+                cell_type_df[COLUMNS.CELL_ID.value].values.tolist(),
+                cell_type_df[COLUMNS.PREDICTION_SCORE.value].values.tolist()
+            )
+        ]
+
+        ancestors_and_sets.append((
+            named_ancestors_reversed,
+            set_value
+        ))
+
+    # Pop off all ancestors that are the same for all cell types.
+    # e.g. 'cell', 'native cell', ...
+    ancestor_list_lens = [len(x[0]) for x in ancestors_and_sets]
+    min_ancestor_list_len = min(ancestor_list_lens)
+    assert min_ancestor_list_len >= 1
+    for level in range(min_ancestor_list_len - 1):
+        unique_level_cell_types = set()
+        for ancestors, cell_set in ancestors_and_sets:
+            unique_level_cell_types.add(ancestors[0])
+
+        if len(unique_level_cell_types) == 1:
+            for ancestors, cell_set in ancestors_and_sets:
+                ancestors.pop(0)
+        else:
+            break
+
+    # Construct a hierarchy of cell types.
+    def find_or_create_parent(d, keys, child):
+        key = keys[0]
+
+        if key in d and isinstance(d[key], dict):
```

Comment:
Would a defaultdict be useful here?

Reply:
I am not very familiar with defaultdict, but the condition and the `result = ` side effect seem more complex than the typical `list`, `set`, and `int` examples that I see in the python docs about it.